File: app/src/main/java/com/medo/tweetspie/main/MainPresenter.java
package com.medo.tweetspie.main;

import com.medo.tweetspie.base.AbsViewPresenter;
import com.medo.tweetspie.database.RealmInteractor;
import com.medo.tweetspie.database.model.RealmTweet;
import com.medo.tweetspie.system.PreferencesProvider;

import io.realm.OrderedRealmCollection;


class MainPresenter extends AbsViewPresenter<MainContract.View>
        implements MainContract.Presenter {

  private final PreferencesProvider preferences;
  private final RealmInteractor realmInteractor;

  MainPresenter(PreferencesProvider preferencesProvider, RealmInteractor realmInteractor) {

    this.preferences = preferencesProvider;
    this.realmInteractor = realmInteractor;
  }

  @Override
  public void onAttach(MainContract.View view) {

    super.onAttach(view);
    if (!preferences.has(PreferencesProvider.USERNAME)) {
      view.startOnboarding();
    }
    else {
      view.initUi();
      view.loadData();
      // show the persisted tweets if any
      OrderedRealmCollection<RealmTweet> tweets = realmInteractor.getTweets();
      if (!tweets.isEmpty()) {
        view.showData(tweets, true);
      }
    }
  }

  @Override
  public void onDetach() {

    super.onDetach();
    realmInteractor.onDestroy();
  }

  @Override
  public void onOnboardingSuccess() {

    if (view != null) {
      view.initUi();
      view.loadData();
    }
  }

  @Override
  public void onOnboardingFailure() {

    if (view != null) {
      view.exit();
    }
  }

  @Override
  public void onDataLoaded(boolean success) {

    if (view != null) {
      if (success) {
        OrderedRealmCollection<RealmTweet> tweets = realmInteractor.getTweets();
        view.showData(tweets, false);
      }
      else {
        // TODO show error
        view.showError();
      }
    }
  }
}

File: app/src/main/java/com/medo/tweetspie/main/MainPresenter.kt
package com.medo.tweetspie.main

import com.medo.tweetspie.R
import com.medo.tweetspie.base.AbsBasePresenter
import com.medo.tweetspie.system.Preferences
import com.medo.tweetspie.system.Resources
import com.medo.tweetspie.system.USERNAME

class MainPresenter(
        private val preferences: Preferences,
        private val resources: Resources
) : AbsBasePresenter<MainContract.View>(), MainContract.Presenter {

    override fun onAttach(view: MainContract.View) {
        super.onAttach(view)
        if (!preferences.has(USERNAME)) view.startOnboarding()
        else view.loadData()
    }

    override fun onOnboardingSuccess() {
        view?.loadData()
    }

    override fun onOnboardingFailure() {
        view?.exit()
    }

    override fun onDataLoaded(success: Boolean) {
        if (success) view?.showData(emptyList(), false)
        else view?.showError(resources.getString(R.string.error_load_data_failure))
    }
}